pkg/promutil/migrate.go

```diff
@@ -64,63 +63,76 @@ func BuildMetrics(results []model.CloudwatchMetricResult, labelsSnakeCase bool,
 	for _, result := range results {
 		contextLabels := contextToLabels(result.Context, labelsSnakeCase, logger)
 		for _, metric := range result.Data {
-			for _, statistic := range metric.Statistics {
-				var includeTimestamp bool
-				if metric.AddCloudwatchTimestamp != nil {
-					includeTimestamp = *metric.AddCloudwatchTimestamp
-				}
-				exportedDatapoint, timestamp, err := getDatapoint(metric, statistic)
-				if err != nil {
-					return nil, nil, err
-				}
-				if exportedDatapoint == nil && (metric.AddCloudwatchTimestamp == nil || !*metric.AddCloudwatchTimestamp) {
-					exportedDatapoint = aws.Float64(math.NaN())
-					includeTimestamp = false
-					if *metric.NilToZero {
-						exportedDatapoint = aws.Float64(0)
+			var statisticsRepresentedInMetric []string
+			includeTimestamp := false
+			if metric.MetricConfig.AddCloudwatchTimestamp != nil {
+				includeTimestamp = *metric.MetricConfig.AddCloudwatchTimestamp
+			}
+			// A result from GetMetricData is for a single statistic noted on the result while a
+			// result from GetMetricStatistics is for all statistics configured for the metric
+			if metric.GetMetricDataResult != nil {
+				statisticsRepresentedInMetric = []string{metric.GetMetricDataResult.Statistic}
+			} else {
+				statisticsRepresentedInMetric = metric.MetricConfig.Statistics
+			}
+			for _, statistic := range statisticsRepresentedInMetric {
+				var exportedDatapoint float64
+				var timestamp time.Time
+				if metric.GetMetricDataResult != nil {
+					exportedDatapoint = metric.GetMetricDataResult.Datapoint
+					timestamp = metric.GetMetricDataResult.Timestamp
+				} else {
+					dataPoint, ts, err := getDatapoint(metric.MetricConfig.Name, metric.GetMetricStatisticResult, statistic)
+					if err != nil {
+						return nil, nil, err
+					}
+					timestamp = ts
+					if dataPoint == nil {
+						exportedDatapoint = math.NaN()
+						// If there was no datapoint then timestamp is a default value
+						includeTimestamp = false
 					}
 				}
+				if *metric.MetricConfig.NilToZero && math.IsNaN(exportedDatapoint) {
+					exportedDatapoint = 0
+				}
```

Comment:
I think this change fixes a bug which exists with `nilToZero`. A discovery job can never have a `GetMetricDataResult` with a nil DataPoint. When we map the data from AWS `Datapoint` is one of the only fields [which is not a pointer](https://github.com/kgeckhart/yet-another-cloudwatch-exporter/blob/c4ddf3cc6c0ba5d08510dd992443b5ea25ce7a8b/pkg/clients/cloudwatch/client.go#L45-L49). After that we [remove all metrics from the result](https://github.com/kgeckhart/yet-another-cloudwatch-exporter/blob/c4ddf3cc6c0ba5d08510dd992443b5ea25ce7a8b/pkg/job/discovery.go#L109-L114) which didn't have a datapoint mapped. 

Based on this I think the old code is completely broken and could never work for a discovery job,

```go
if exportedDatapoint == nil && (metric.AddCloudwatchTimestamp == nil || !*metric.AddCloudwatchTimestamp) {
    exportedDatapoint = aws.Float64(math.NaN())
    includeTimestamp = false
    if *metric.NilToZero {
        exportedDatapoint = aws.Float64(0)
```
We only use `nilToZero` when the actual value is nil (can never happen) and some extra conditionals related if the CloudWatch timestamp is to be used. I don't understand how those two conditions are related at all?

The new logic here splits this up so that we properly handle the valid nil data point case for a static job and all jobs can have `NaN` converted to `0` according to `NilToZero`.

Reply:
yhea that logic above is really weird (like the mix of checking if there's data and the CW ts should be used). +1 on dropping that in favour of yours

Reply:
> A discovery job can never have a GetMetricDataResult with a nil DataPoint.

Before the change where we stopped using pointers, it could have been nil in the case the CW returns the metric without any data point (I think it's the case that @nhinds refers to in the other comment).